test/subspace-test-service/src/lib.rs

```diff
@@ -464,7 +464,10 @@ impl MockConsensusNode {
             self.acknowledgement_sender_subscribers
                 .retain(|subscriber| {
                     subscriber
+                        // Send 2 `acknowledgement_sender` to ensure all the domain worker's previous
+                        // tasks have finished
                         .unbounded_send(acknowledgement_sender.clone())
+                        .and_then(|_| subscriber.unbounded_send(acknowledgement_sender.clone()))
```

Comment:
I do not understand this. How exactly is it ensuring something? All I see is two in-flight messages instead of one.

Reply:
Not really in-flight message, the message is an `acknowledgement_sender` and this function will wait until all the `acknowledgement_sender ` respond before returning.

Each message will result in one round of iteration of the loop and enter this branch here:
https://github.com/subspace/subspace/blob/bab56959e78adb13ef3da54ac9837c44ee14fa3d/domains/client/domain-operator/src/domain_worker_starter.rs#L208

The message is handled in the same loop as the slot notification, thus as long as we can ensure the slot notification is sent before the `acknowledgement_sender` thus handled before the `acknowledgement_sender` we can ensure the slot notification is finished processed when the `acknowledgement_sender` responds, we have ensure this in main by:
https://github.com/subspace/subspace/blob/bab56959e78adb13ef3da54ac9837c44ee14fa3d/test/subspace-test-service/src/lib.rs#L417-L421
but it does not work sometimes as the log shows.

I also suspect it is due to https://github.com/smol-rs/async-channel/issues/54 trying to send 2 `acknowledgement_sender` thus polling the `new_slot_notification_stream` twice, and it seems to work. I have changed the `test_domain_block_production` to produce 500~1000 blocks instead of 50 to reproduce the issue locally, without this change the issue can be reproduced about 1 out of 5 runs, but never reproduce with this change.

Reply:
This is really hard to follow. Not only sending a sender twice feels like a hack for architectural issue, this whole method is called incorrectly, it should probably be called `wait_for_confirm_acknowledgement` and while at it, it should be a sync method that returns a future, so you can subscribe to acknowledgements first, then trigger the thing that will do acknowledgement and then wait for it.

In particular (without reading code carefully) I suspect you need to subscribe first, then call `subscriber.unbounded_send(value)` and only then wait for acknowledgements to arrive. That way to fundamentally avoid race condition, current implementation though might just decrease probability of the issue happening.